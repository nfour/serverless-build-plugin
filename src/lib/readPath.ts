import { map } from 'bluebird';
import { readdir, stat, Stats } from 'fs-extra';

export interface IFileParams {
  filePath: string;
  stats: Stats;
  startPath: string;
  previousPaths: string[];
}

export async function readPath (startDirectory: string, inputParams: {
  /** Depth at while to stop traversal */
  depth?: number;

  onFileFilter? (params: IFileParams): boolean|Promise<boolean>;
  onFile (params: IFileParams);
}) {

  type ITraverseParams = Parameters<typeof readPath>[1] & { startPath: string, previousPaths: string[] };

  async function traversePath ({  }: ITraverseParams) {
    const paths = await readdir(startDirectory);

    console.log({ paths });

    await map(paths, (path) => {
      return traversePath(path, { onFileFilter, onFile, depth: depth - 1 });
    });
  }

  return traversePath(startDirectory, { });
}
