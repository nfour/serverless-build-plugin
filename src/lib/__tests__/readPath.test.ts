import { resolve } from 'path';

import { readPath } from '../readPath';

const testDir = resolve(__dirname, '../../__tests__/project');

it('reads regular file nested in 1 folder', async () => {
  const files: string[] = [];

  await readPath(testDir, {
    depth: 2,
    async onFile ({ startPath, previousPaths, filePath, stats }) {
      console.log({ previousPaths });
      files.push(filePath);
    },
  });

  expect(files.sort()).toMatchSnapshot();
});

it('can filter out paths matching /c/', async () => {
  const files: string[] = [];

  await readPath(testDir, {
    depth: 4,
    async onFile ({ startPath, previousPaths, filePath, stats }) {
      if (!/\/c\//.test(filePath)) { return true; }

      files.push(filePath);
    },
  });

  expect(files.sort()).toMatchSnapshot();
});

it('can traverse circular symlinks up to depth', async () => {
  const files: string[] = [];

  await readPath(testDir, {
    depth: 20,
    async onFile ({ startPath, previousPaths, filePath, stats }) {
      files.push(filePath);
    },
  });

  expect(files.sort()).toMatchSnapshot();
});
