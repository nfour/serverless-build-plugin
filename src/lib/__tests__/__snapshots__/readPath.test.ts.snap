// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can filter out paths matching /c/ 1`] = `
Array [
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/c/projectSym/a",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/c/projectSym/b/d",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/c/projectSym/bSym/d",
]
`;

exports[`can traverse circular symlinks up to depth 1`] = `
Array [
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/a",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/b/d",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/bSym/d",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/c/projectSym/a",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/c/projectSym/b/d",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/c/projectSym/bSym/d",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/c/projectSym/c/projectSym/a",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/c/projectSym/c/projectSym/b/d",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/c/projectSym/c/projectSym/bSym/d",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/c/projectSym/c/projectSym/c/projectSym/a",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/c/projectSym/c/projectSym/c/projectSym/b/d",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/c/projectSym/c/projectSym/c/projectSym/bSym/d",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/c/projectSym/c/projectSym/c/projectSym/c/projectSym/a",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/c/projectSym/c/projectSym/c/projectSym/c/projectSym/b/d",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/c/projectSym/c/projectSym/c/projectSym/c/projectSym/bSym/d",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/a",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/b/d",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/bSym/d",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/a",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/b/d",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/bSym/d",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/a",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/b/d",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/bSym/d",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/a",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/b/d",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/bSym/d",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/a",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/b/d",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/c/projectSym/bSym/d",
]
`;

exports[`reads regular file nested in 1 folder 1`] = `
Array [
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/a",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/b/d",
  "/home/sam/W/serverless-build-plugin/src/__tests__/project/bSym/d",
]
`;
